e2e/scripts/mrt-target-manager.js

```diff
@@ -0,0 +1,322 @@
+            roleArn: process.env.CI ? options.roleArn : options.roleArn, // Don't use role ARN in CI since AWS credentials action handles it
+            roleSessionName: options.roleSessionName || 'LocalDev'
+        })
+    }
+
+    async initialize() {
+        await this.s3Client.initialize()
+    }
+
+    /**
+     * Convert a ReadableStream object returned by the S3 client to a string
+     * @param {Stream} stream - The stream to convert
+     * @returns {Promise<string>} - The string representation of the stream
+     */
+    async streamToString(stream) {
+        const chunks = []
+        for await (const chunk of stream) {
+            chunks.push(chunk)
+        }
+        return Buffer.concat(chunks).toString()
+    }
+
+    /**
+     * Download the pool file and return data with ETag
+     */
+    async downloadPoolFile() {
+        try {
+            const downloadResult = await this.s3Client.download(this.bucket, this.poolDataFileKey)
+            // Convert stream to string and parse JSON
+            const contentString = await this.streamToString(downloadResult.body)
+            const poolData = JSON.parse(contentString)
+
+            return {
+                ...downloadResult,
+                poolData
+            }
+        } catch (error) {
+            if (error.name === 'NoSuchKey') {
+                console.log('❌ Pool file not found.')
+            }
+            throw error
+        }
+    }
+
+    /**
+     * Find an available environment of the specified type
+     */
+    findAvailableEnvironment(poolData) {
+        const availableEnvs = poolData.environments.filter(
+            (env) => env.ciAvailability === 'available'
+        )
+
+        if (availableEnvs.length === 0) {
+            return null
+        }
+
+        return availableEnvs[0]
+    }
+
+    /**
+     * Mark environment as in-use by current PR
+     */
+    updateMRTTargetStatus(poolData, environment, ciAvailability) {
+        const updatedPoolData = {
+            ...poolData,
+            environments: poolData.environments.map((env) => {
+                if (env.slug === environment.slug) {
+                    const updatedEnv = {
+                        ...env,
+                        ciAvailability,
+                        ciAcquiredAt: new Date().toISOString(),
+                        ciLastUsed: new Date().toISOString()
+                    }
+
+                    if (ciAvailability === 'in-use') {
+                        if (this.prNumber) updatedEnv.ciPRNumber = this.prNumber
+                        if (this.branch) updatedEnv.ciBranch = this.branch
+                        if (this.runId) updatedEnv.ciRunId = this.runId
+                    } else if (ciAvailability === 'available') {
+                        delete updatedEnv.ciPRNumber
+                        delete updatedEnv.ciBranch
+                        delete updatedEnv.ciRunId
+                        delete updatedEnv.ciAcquiredAt
```

Comment:
if all ci info is grouped, we can simply `delete updatedEnv.ci`

Reply:
Ahh! Good catch. Makes sense :D